adapters/operaads/operaads.go

```diff
@@ -0,0 +1,233 @@
+package operaads
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/prebid/prebid-server/macros"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	epTemplate *template.Template
+}
+
+// Builder builds a new instance of the operaads adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	epTemplate, err := template.New("endpoint").Parse(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	bidder := &adapter{
+		epTemplate: epTemplate,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	impCount := len(request.Imp)
+	requestData := make([]*adapters.RequestData, 0, impCount)
+	errs := []error{}
+	request, err := deepCopyRequest(request)
+	if err != nil {
+		errs = append(errs, &errortypes.BadInput{
+			Message: err.Error(),
+		})
+		return nil, errs
+	}
```

Comment:
@yuanjize I see what you're trying to do here but I think this is a bit of overkill. I recommend removing the deep copy logic and instead considering my suggestion in `convertImpression` below.